### PR TITLE
Updating users

### PR DESCRIPTION
**Pull request by kalpeshDocFox**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -11,11 +11,12 @@ def can_authenticate(user)
   def create
     @user = User.find_by(email: params.dig(:session, :email)&.downcase)
     if can_authenticate(@user)
```
Comment (ldmaselela) [inline app/controllers/sessions_controller.rb]: Shouldn't can_authenticate return a boolean value ? if so I think you need to change it to can_authenticate?(@user)

```suggestion
    if can_authenticate?(@user)
```

File: test/controllers/users_controller_test.rb [hunk 2]
```diff
@@ -1,9 +1,49 @@
...
+  test "PATCH should redirect edit when not logged in" do
+    get edit_user_path(@user)
+    assert_not flash.empty?
+    assert_redirected_to login_url
+  end
+
+  test "should not allow the admin attribute to be edited via the web" do
+    log_in_as(@other_user)
+    assert_not @other_user.admin?
+    patch user_path(@other_user), params: {
```
Comment (ldmaselela) [inline test/controllers/users_controller_test.rb]: The indentation on the params hag doesn't seem to be correct, maybe consider moving params to a method?

File: test/controllers/users_controller_test.rb [hunk 3]
```diff
@@ -1,9 +1,49 @@
...
+                                    user: { password: "password",
+                                            password_confirmation: "password",
+                                            admin: true } }
+    assert_not @other_user.admin?                              
+  end
+
+  test "PATCH should redirect update when not logged in" do
+    patch user_path(@user), params: { user: { name: @user.name,
```
Comment (ldmaselela) [inline test/controllers/users_controller_test.rb]: tiddo on indentation

File: db/seeds.rb [hunk 4]
```diff
@@ -1,7 +1,17 @@
...
-# The data can then be loaded with the rails db:seed command (or created alongside the database with db:setup).
-#
-# Examples:
-#
-#   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
-#   Character.create(name: 'Luke', movie: movies.first)
+# Create a main sample user.
+User.create!(name:  "Example User",
+    email: "example@railstutorial.org",
+    password:              "foobar",
```
Comment (ldmaselela) [inline db/seeds.rb]: You don't need the white space 
```suggestion
    password:"foobar",
```

File: app/controllers/sessions_controller.rb (hunk 1 above)
Comment (c13e) [reply app/controllers/sessions_controller.rb]: yup, common convention in Ruby is to add a question mark at the end to signal that it returns a boolean :)

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -1,4 +1,12 @@
 class UsersController < ApplicationController
+  before_action :logged_in_user, only: [:index, :edit, :update, :destroy]
+  before_action :correct_user, only: [:edit, :update]
+  before_action :admin_user, only: :destroy
+  
+  def index
+      @users = User.paginate(page: params[:page])
```
Comment (c13e) [inline app/controllers/users_controller.rb]:
```suggestion
    @users = User.paginate(page: params[:page]) 
```
as previously mentioned, try to stick to 2-space indentation, I'd strongly suggest setting this in you text editor so we don't always have to give the same feedback in future PRs

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -18,6 +36,36 @@ def create
...
+      redirect_to login_url
+    end 
+  end
+
+  # Confirms an admin user. 
+  def admin_user
+    redirect_to(root_url) unless current_user.admin? 
+  end
+
+# Confirms the correct user.
```
Comment (c13e) [inline app/controllers/users_controller.rb]:
```suggestion
  # Confirms the correct user.
```

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -18,6 +36,36 @@ def create
...
+    @user = User.find(params[:id]) 
+    redirect_to(root_url) unless @user == current_user
+  end
+
+  def update
+    @user = User.find(params[:id])
+    if @user.update(user_params)
+      flash[:success] = "Profile updated"
+      redirect_to @user
+  else
```
Comment (c13e) [inline app/controllers/users_controller.rb]:
```suggestion
    else
```

File: app/models/user.rb [hunk 8]
```diff
@@ -14,9 +14,10 @@ class User < ApplicationRecord
             format: { with: VALID_EMAIL_REGEX },
             uniqueness: { case_sensitive: false }
   has_secure_password
+  
   validates :password,
             presence: true,
-            length: { minimum: PASSSWORD_MIN_LENGTH }
+            length: { minimum: PASSSWORD_MIN_LENGTH }, allow_nil: true
```
Comment (c13e) [inline app/models/user.rb]:
```suggestion
            length: { minimum: PASSSWORD_MIN_LENGTH },
            allow_nil: true
```

File: app/views/users/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,13 @@
+<% provide(:title, 'All users') %> 
+<h1>All users</h1>
+
+
+<%= will_paginate %>
+
+<ul class="users">
+  <% @users.each do |user| %>
+    <%= render user %> 
+    <% end %>
```
Comment (c13e) [inline app/views/users/index.html.erb]:
```suggestion
  <% @users.each do |user| %>
    <%= render user %> 
  <% end %>
```

File: db/seeds.rb [hunk 10]
```diff
@@ -1,7 +1,17 @@
...
+# Generate a bunch of additional users.
+99.times do |n|
+name = Faker::Name.name
+email = "example-#{n+1}@railstutorial.org" 
+password = "password"
+User.create!(name: name,
+            email: email,
+            password:password,
+            password_confirmation: password)
+end
```
Comment (c13e) [inline db/seeds.rb]:
```suggestion
99.times do |n|
  name = Faker::Name.name
  email = "example-#{n+1}@railstutorial.org" 
  password = "password"
  User.create!(
    name: name,
    email: email,
    password:password,
    password_confirmation: password
  )
end
```

**Review by c13e — APPROVED** [review]
Noice!

**Review by ldmaselela — APPROVED** [review]

**Pull request merged**